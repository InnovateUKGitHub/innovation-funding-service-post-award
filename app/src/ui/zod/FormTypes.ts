enum FormTypes {
  // MSP Document Share page
  ProjectLevelDelete = "projectLevelDelete",
  PartnerLevelDelete = "partnerLevelDelete",
  ProjectLevelUpload = "projectLevelUpload",

  // Claim page
  ClaimLevelDelete = "claimLevelDelete",
  ClaimLevelUpload = "claimLevelUpload",
  ClaimForecastSaveAndContinue = "claimForecastSaveAndContinue",
  ClaimForecastSaveAndQuit = "claimForecastSaveAndQuit",
  ClaimLineItemSaveAndDocuments = "claimLineItemSaveAndDocuments",
  ClaimLineItemSaveAndQuit = "claimLineItemSaveAndQuit",

  // Project Setup
  ProjectSetupBankDetails = "projectSetupBankDetails",
  ProjectSetupContactAssociate = "projectSetupContactAssociate",
  ProjectSetupForecast = "projectSetupForecast",

  // Forecast Tile
  ForecastTileForecast = "forecastTileForecast",

  // Claim Details page
  ClaimDetailLevelUpload = "claimDetailLevelUpload",
  ClaimDetailLevelDelete = "claimDetailLevelDelete",

  // Claim review page
  ClaimReviewLevelSaveAndContinue = "claimReviewSaveAndContinue",
  ClaimReviewLevelUpload = "claimReviewLevelUpload",
  ClaimReviewLevelDelete = "claimReviewLevelDelete",

  // Project Change Requests
  ProjectChangeRequestCreate = "projectChangeRequestCreate",
  ProjectChangeRequestUpdateTypes = "projectChangeRequestUpdateTypes",

  PcrLevelDelete = "pcrLevelDelete",
  PcrLevelUpload = "pcrLevelUpload",

  PcrFinancialVirementsSummary = "pcrFinancialVirementsSummary",
  PcrFinancialVirementsCostCategorySaveAndContinue = "pcrFinancialVirementsCostCategorySaveAndContinue",
  PcrReallocateCostsChangeRemainingGrant = "pcrReallocateCostsChangeRemainingGrant",

  PcrApproveNewSubcontractorStep = "pcrApproveNewSubcontractorDetails",
  PcrApproveNewSubcontractorSummary = "pcrApproveNewSubcontractorSummary",

  PcrAddPartnerCompaniesHouseStepSaveAndContinue = "pcrAddPartnerCompaniesHouseStepSaveAndContinue",
  PcrAddPartnerCompaniesHouseStepSaveAndQuit = "pcrAddPartnerCompaniesHouseStepSaveAndQuit",

  PcrChangeProjectScopeProposedPublicDescriptionStepSaveAndContinue = "pcrChangeProjectScopeProposedPublicDescriptionStepSaveAndContinue",
  PcrChangeProjectScopeProposedProjectSummaryStepSaveAndContinue = "pcrChangeProjectScopeProposedProjectSummaryStepSaveAndContinue",
  PcrChangeProjectScopeSummary = "pcrChangeProjectScopeSummary",

  PcrRenamePartnerStep = "pcrRenamePartnerStep",
  PcrRenamePartnerFilesStep = "pcrRenamePartnerFilesStep",
  PcrRenamePartnerSummary = "pcrRenamePartnerSummary",

  PcrPrepareReasoningStep = "pcrPrepareReasoningStep",
  PcrPrepareReasoningFilesStep = "pcrPrepareReasoningFilesStep",
  PcrPrepareReasoningSummary = "pcrPrepareReasoningSummary",

  LoanLevelUpload = "loanLevelUpload",
  LoanLevelDelete = "loanLevelDelete",
  LoanRequest = "loanRequest",
}

export { FormTypes };
