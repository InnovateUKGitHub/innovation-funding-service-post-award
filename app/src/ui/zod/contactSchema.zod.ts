import { z } from "zod";
import { dateValidation, pclIdValidation } from "./helperValidators/helperValidators.zod";
import { makeZodI18nMap } from "@shared/zodi18n";

const contactSchemaErrorMap = makeZodI18nMap({ keyPrefix: ["contact"] });

const contactDtoSchema = z.object({
  id: pclIdValidation,
  associateStartDate: dateValidation,
});

const multipleContactDtoSchema = z.object({
  contacts: z.array(contactDtoSchema),
});

type MultipleContactDtoSchemaType = typeof multipleContactDtoSchema;
type ContactDtoSchemaType = typeof contactDtoSchema;

export {
  contactDtoSchema,
  multipleContactDtoSchema,
  contactSchemaErrorMap,
  ContactDtoSchemaType,
  MultipleContactDtoSchemaType,
};
