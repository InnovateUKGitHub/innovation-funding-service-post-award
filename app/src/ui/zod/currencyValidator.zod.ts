import { parseCurrency, validCurrencyRegex } from "@framework/util/numberHelper";
import { isNil } from "lodash";
import { z, ZodIssueCode } from "zod";

const defaultMaxCurrency = 999_999_999_999;
const defaultMinCurrency = 0;

export const getGenericCurrencyValidation = ({
  label,
  max = defaultMaxCurrency,
  min = defaultMinCurrency,
  required = false,
}: {
  label: string;
  max?: number;
  min?: number;
  required?: boolean;
}) => {
  return z
    .string()
    .nullable()
    .superRefine((val, ctx) => {
      const currency = parseCurrency(val);
      const isEmpty = (typeof val === "string" && val.trim() === "") || isNil(val);
      if (required && isEmpty) {
        return ctx.addIssue({
          code: ZodIssueCode.custom,
          params: {
            label,
            generic: ["currency"],
            i18n: "errors.required",
          },
        });
      }

      if (!isEmpty) {
        if (/[\$€¥]+/.test(val)) {
          return ctx.addIssue({
            code: ZodIssueCode.custom,
            params: {
              label,
              generic: ["currency"],
              i18n: "errors.not_pounds",
            },
          });
        }
        // Check if the string can even be parsed
        if (isNaN(currency)) {
          return ctx.addIssue({
            code: ZodIssueCode.custom,
            params: {
              label,
              generic: ["currency"],
              i18n: "errors.not_a_number",
            },
          });
        }

        // Make sure our currency isn't so big as to break our server
        if (currency > max) {
          return ctx.addIssue({
            code: ZodIssueCode.custom,
            params: {
              label,
              count: max,
              generic: ["currency"],
              i18n: "errors.too_big",
            },
          });
        }

        if (currency < min) {
          return ctx.addIssue({
            code: ZodIssueCode.custom,
            params: {
              label,
              count: min,
              generic: ["currency"],
              i18n: "errors.too_small",
            },
          });
        }
        if (/\.\d\d\d/.test(val)) {
          return ctx.addIssue({
            code: ZodIssueCode.custom,
            params: {
              label,
              generic: ["currency"],
              i18n: "errors.two_decimal_places",
            },
          });
        }

        if (!validCurrencyRegex.test(val)) {
          return ctx.addIssue({
            code: ZodIssueCode.custom,
            params: {
              label,
              generic: ["currency"],
              i18n: "errors.invalid_currency",
            },
          });
        }
      }
    })
    .transform(x => {
      if ((typeof x === "string" && x.trim() === "") || isNil(x)) {
        return null;
      }
      return parseCurrency(x);
    });
};
