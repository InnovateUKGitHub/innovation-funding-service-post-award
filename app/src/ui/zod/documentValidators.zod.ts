import {
  allowedClaimDocuments,
  allowedImpactManagementClaimDocuments,
  allowedProjectLevelDocuments,
  DocumentDescription,
  allowedPcrLevelDocuments,
} from "@framework/constants/documentDescription";
import { IAppOptions } from "@framework/types/IAppOptions";
import { z } from "zod";
import { FormTypes } from "./FormTypes";
import {
  projectIdValidation,
  emptyStringToUndefinedValidation,
  partnerIdValidation,
  getMultiFileValidation,
  periodIdValidation,
  pcrItemIdValidation,
  costCategoryIdValidation,
  pcrIdValidation,
} from "./helperValidators.zod";
import { ProjectDto } from "@framework/dtos/projectDto";
import { ImpactManagementParticipation } from "@framework/constants/competitionTypes";
import { makeZodI18nMap } from "@shared/zodi18n";

type ProjectLevelUploadSchemaType = ReturnType<typeof getProjectLevelUpload>;
const getProjectLevelUpload = (config: IAppOptions) =>
  z.object({
    form: z.literal(FormTypes.ProjectLevelUpload),
    projectId: projectIdValidation,
    partnerId: z.union([emptyStringToUndefinedValidation, partnerIdValidation]),
    description: z.union([
      emptyStringToUndefinedValidation,
      z.coerce
        .number()
        .refine(x => allowedProjectLevelDocuments.includes(x))
        .optional()
        .transform(x => x as DocumentDescription),
    ]),
    files: getMultiFileValidation(config),
  });

type UploadBankStatementSchemaType = ReturnType<typeof getBankStatementUpload>;
const getBankStatementUpload = (config: IAppOptions) =>
  z.object({
    form: z.literal(FormTypes.ProjectLevelUpload),
    projectId: projectIdValidation,
    partnerId: z.union([emptyStringToUndefinedValidation, partnerIdValidation]),
    description: z.union([
      emptyStringToUndefinedValidation,
      z.coerce
        .number()
        .refine(x => [DocumentDescription.BankStatement].includes(x))
        .optional()
        .transform(x => x as DocumentDescription),
    ]),
    files: getMultiFileValidation(config),
  });

interface ClaimLevelUploadSchemaExtraProps {
  config: IAppOptions;
  project: Pick<ProjectDto, "impactManagementParticipation">;
}

type ClaimLevelUploadSchemaType = ReturnType<typeof getClaimLevelUpload>;
const documentsErrorMap = makeZodI18nMap({ keyPrefix: ["documents"] });
const getClaimLevelUpload = ({ config, project }: ClaimLevelUploadSchemaExtraProps) =>
  z.object({
    form: z.union([z.literal(FormTypes.ClaimLevelUpload), z.literal(FormTypes.ClaimReviewLevelUpload)]),
    projectId: projectIdValidation,
    partnerId: partnerIdValidation,
    description: z.union([
      emptyStringToUndefinedValidation,
      z.coerce
        .number()
        .refine(x =>
          (project.impactManagementParticipation === ImpactManagementParticipation.Yes
            ? allowedImpactManagementClaimDocuments
            : allowedClaimDocuments
          ).includes(x),
        )
        .optional()
        .transform(x => x as DocumentDescription),
    ]),
    periodId: periodIdValidation,
    files: getMultiFileValidation(config),
  });

type ClaimDetailLevelUploadSchemaType = ReturnType<typeof getClaimDetailLevelUpload>;
const getClaimDetailLevelUpload = ({ config, project }: ClaimLevelUploadSchemaExtraProps) =>
  z.object({
    form: z.literal(FormTypes.ClaimDetailLevelUpload),
    projectId: projectIdValidation,
    partnerId: partnerIdValidation,
    costCategoryId: costCategoryIdValidation,
    description: z.union([
      emptyStringToUndefinedValidation,
      z.coerce
        .number()
        .refine(x =>
          (project.impactManagementParticipation === ImpactManagementParticipation.Yes
            ? allowedImpactManagementClaimDocuments
            : allowedClaimDocuments
          ).includes(x),
        )
        .optional()
        .transform(x => x as DocumentDescription),
    ]),
    periodId: periodIdValidation,
    files: getMultiFileValidation(config),
  });

const getPcrLevelUpload = ({ config }: { config: IAppOptions }) =>
  z.object({
    form: z.literal(FormTypes.PcrLevelUpload),
    projectId: projectIdValidation,
    projectChangeRequestIdOrItemId: pcrItemIdValidation,
    description: z.union([
      emptyStringToUndefinedValidation,
      z.coerce
        .number()
        .refine(x => allowedPcrLevelDocuments.includes(x))
        .optional()
        .transform(x => x as DocumentDescription),
    ]),
    files: getMultiFileValidation(config),
  });

type PcrLevelUploadSchemaType = ReturnType<typeof getPcrLevelUpload>;

const projectLevelDelete = z.object({
  form: z.literal(FormTypes.ProjectLevelDelete),
  projectId: projectIdValidation,
  documentId: z.string(),
});

const partnerLevelDelete = z.object({
  form: z.literal(FormTypes.PartnerLevelDelete),
  projectId: projectIdValidation,
  partnerId: partnerIdValidation,
  documentId: z.string(),
});

type ClaimLevelDeleteSchemaType = typeof claimLevelDelete;
const claimLevelDelete = z.object({
  form: z.union([z.literal(FormTypes.ClaimLevelDelete), z.literal(FormTypes.ClaimReviewLevelDelete)]),
  projectId: projectIdValidation,
  partnerId: partnerIdValidation,
  periodId: periodIdValidation,
  documentId: z.string(),
});

const claimDetailLevelDelete = z.object({
  form: z.literal(FormTypes.ClaimDetailLevelDelete),
  projectId: projectIdValidation,
  partnerId: partnerIdValidation,
  periodId: periodIdValidation,
  costCategoryId: costCategoryIdValidation,
  documentId: z.string(),
});

const pcrLevelDelete = z.object({
  form: z.literal(FormTypes.PcrLevelDelete),
  projectId: projectIdValidation,
  projectChangeRequestIdOrItemId: z.union([pcrItemIdValidation, pcrIdValidation]),
  documentId: z.string(),
});

const projectOrPartnerLevelDelete = z.discriminatedUnion("form", [partnerLevelDelete, projectLevelDelete]);

export {
  getPcrLevelUpload,
  getProjectLevelUpload,
  projectLevelDelete,
  partnerLevelDelete,
  getClaimLevelUpload,
  getClaimDetailLevelUpload,
  claimDetailLevelDelete,
  claimLevelDelete,
  pcrLevelDelete,
  projectOrPartnerLevelDelete,
  getBankStatementUpload,
  documentsErrorMap,
};
export type {
  ProjectLevelUploadSchemaType,
  ClaimLevelUploadSchemaType,
  UploadBankStatementSchemaType,
  PcrLevelUploadSchemaType,
  ClaimDetailLevelUploadSchemaType,
  ClaimLevelDeleteSchemaType,
};
