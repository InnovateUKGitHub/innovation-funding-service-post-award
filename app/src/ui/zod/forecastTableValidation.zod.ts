import { makeZodI18nMap } from "@shared/zodi18n";
import {
  ClaimStatusGroup,
  getClaimStatusGroup,
} from "@ui/components/atomicDesign/organisms/forecasts/ForecastTable/getForecastHeaderContent";
import {
  mapToForecastTableDto,
  MapToForecastTableProps,
} from "@ui/components/atomicDesign/organisms/forecasts/ForecastTable/NewForecastTable.logic";
import { z, ZodIssueCode } from "zod";
import { FormTypes } from "./FormTypes";
import {
  booleanValidation,
  partnerIdValidation,
  profileIdValidation,
  projectIdValidation,
} from "./helperValidators.zod";
import { getGenericCurrencyValidation } from "./currencyValidator.zod";

type ForecastTableSchemaType = ReturnType<typeof getForecastTableValidation>["schema"];
const getForecastTableValidation = (data: Omit<MapToForecastTableProps, "clientProfiles">) => {
  const errorMap = makeZodI18nMap({ keyPrefix: ["forecastTable"], context: data });

  const schema = z
    .object({
      projectId: projectIdValidation,
      partnerId: partnerIdValidation,
      form: z.union([
        z.literal(FormTypes.ClaimForecastSaveAndContinue),
        z.literal(FormTypes.ClaimForecastSaveAndQuit),
        z.literal(FormTypes.ProjectSetupForecast),
        z.literal(FormTypes.ForecastTileForecast),
      ]),
      profile: z.record(
        profileIdValidation,
        getGenericCurrencyValidation({
          label: "forms.forecastTable.profile.label",
          min: -1_000_000_000,
          required: true,
        }),
      ),
      submit: booleanValidation,
    })
    .superRefine(({ profile: clientProfiles, form, submit }, { addIssue, path }) => {
      const table = mapToForecastTableDto({ ...data, clientProfiles });
      const nonForecastClaims = data.claimTotalProjectPeriods.filter(
        x => getClaimStatusGroup(x.status) !== ClaimStatusGroup.FORECAST,
      );
      const finalClaim = nonForecastClaims.find(claim => claim.isFinalClaim);

      if (form === FormTypes.ProjectSetupForecast) {
        if (submit) {
          for (const costCategory of table.costCategories) {
            if (costCategory.differentThanAllocatedCosts && !costCategory.isCalculated) {
              addIssue({
                code: ZodIssueCode.custom,
                path: [...path, "costCategory", costCategory.costCategoryId],
                params: {
                  costCategoryName: costCategory.costCategoryName,
                  i18n: "errors.forecasts_different",
                },
              });
            }
          }
        }
      } else {
        // Don't check if the total is too big for project setup spend profile
        // (there is already a check for each cost cat)
        if (table.totalRow.total > table.totalRow.golCost) {
          addIssue({
            code: ZodIssueCode.too_big,
            maximum: table.totalRow.golCost,
            inclusive: false,
            type: "number",
            path: [...path, "total"],
          });
        }
      }

      if (clientProfiles && finalClaim) {
        addIssue({
          code: ZodIssueCode.custom,
          params: {
            i18n: finalClaim.isApproved ? "errors.final_claim_already_approved" : "errors.final_claim",
          },
        });
      }
    });

  return { errorMap, schema };
};

export { getForecastTableValidation };
export type { ForecastTableSchemaType };
