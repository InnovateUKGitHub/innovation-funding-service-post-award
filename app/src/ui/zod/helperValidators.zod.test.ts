import { IFileWrapper } from "@framework/types/fileWrapper";
import { TestConfig } from "@tests/test-utils/testConfig";
import {
  emptyStringToUndefinedValidation,
  evaluateObject,
  getSingleFileValidation,
  partnerIdValidation,
  periodIdValidation,
  projectIdValidation,
} from "./helperValidators.zod";
import { z } from "zod";

describe("helperValidators", () => {
  describe("projectIdValidation", () => {
    test.each([
      ["decline empty string", "", false],
      ["accept project id", "a0E2600000olswLEAQ", true],
      ["decline project participant id", "a0D2600000zXrAIEA0", false],
      ["decline account id", "0012600001amaskAAA", false],
    ])("%s", (name, input, accept) => {
      const parse = projectIdValidation.safeParse(input);
      expect(parse.success).toBe(accept);
      expect(parse as unknown).toMatchSnapshot();
    });
  });

  describe("partnerIdValidation", () => {
    test.each([
      ["decline empty string", "", false],
      ["decline project id", "a0E2600000olswLEAQ", false],
      ["accept project participant id", "a0D2600000zXrAIEA0", true],
      ["decline account id", "0012600001amaskAAA", false],
    ])("%s", (name, input, accept) => {
      const parse = partnerIdValidation.safeParse(input);
      expect(parse.success).toBe(accept);
      expect(parse as unknown).toMatchSnapshot();
    });
  });

  describe("emptyStringToUndefinedValidation", () => {
    test.each([
      ["accept empty string", "", true],
      ["accept undefined", undefined, true],
      ["decline filled string", "hello world!", false],
      ["decline number", 1.43e99, false],
    ])("%s", (name, input, accept) => {
      const parse = emptyStringToUndefinedValidation.safeParse(input);
      expect(parse.success).toBe(accept);
      expect(parse as unknown).toMatchSnapshot();
    });
  });

  describe("periodIdValidation", () => {
    test.each([
      ["decline 0", 0, false],
      ["accept 1", 1, true],
      ["accept 2", 2, true],
      ["accept 3", 3, true],
      ["accept 5", 5, true],
      ["decline decimal", 1.01, false],
      ["accept string int", "4", true],
      ["decline absurd number", 500, false],
      ["decline infinity", Infinity, false],
    ])("%s", (name, input, accept) => {
      const parse = periodIdValidation.safeParse(input);
      expect(parse.success).toBe(accept);
      expect(parse as unknown).toMatchSnapshot();
    });
  });

  describe("getSingleFileValidation", () => {
    const testConfig = new TestConfig();
    const singleFileValidation = getSingleFileValidation(testConfig.options);

    test.each([
      // Valid files
      ["standard file", { fileName: "awoo.png", size: 120 }, true],

      // Bad prefixes
      [".tar.gz", { fileName: "awoo.tar.gz", size: 12442 }, false],
      [".webm", { fileName: "awoo.webm", size: 12442 }, false],
      [".docm", { fileName: "awoo.docm", size: 12442 }, false],

      // Invalid size files
      ["zero sized file", { fileName: "awoo.png", size: 0 }, false],
      ["very big sized file", { fileName: "awoo.png", size: 908172862 }, false],

      // Invalid name files
      ["no prefix file", { fileName: "jjajangmyeon", size: 4353 }, false],
      ["Linux/UNIX hidden file", { fileName: ".gitignore", size: 4353 }, false],
      ["no basename file", { fileName: ".pptx", size: 4353 }, false],
      ["crazy characters file", { fileName: '*&"Â£&.pptx', size: 4353 }, false],
    ])("%s", (name, input: IFileWrapper, accept) => {
      const parse = singleFileValidation.safeParse(input);
      expect(parse.success).toBe(accept);
      expect(parse as unknown).toMatchSnapshot();
    });
  });

  describe("evaluateObject", () => {
    test.each([
      ["Valid 'marked as complete'", { something: "APPLE", markedAsComplete: true }, true],
      ["Invalid 'marked as complete'", { something: "APPLE", markedAsComplete: false }, false],
      ["Invalid 'NOT marked as complete'", { something: "BANANA", markedAsComplete: true }, false],
      ["Valid 'NOT marked as complete'", { something: "BANANA", markedAsComplete: false }, true],
      ["Invalid", { something: "DURIAN", markedAsComplete: true }, false],
      ["Mal", { something: "DURIAN", markedAsComplete: false }, false],
    ])("%s", (name, objectToValidate, accept) => {
      type ValidationData = { markedAsComplete: boolean; something: string };

      const schema = evaluateObject((data: ValidationData) => ({
        something: data.markedAsComplete ? z.literal("APPLE") : z.literal("BANANA"),
        markedAsComplete: z.boolean(),
      }));

      const markedAsCompleteResult = schema.safeParse(objectToValidate);

      expect(markedAsCompleteResult.success).toBe(accept);
      expect(markedAsCompleteResult as unknown).toMatchSnapshot();
    });
  });
});
