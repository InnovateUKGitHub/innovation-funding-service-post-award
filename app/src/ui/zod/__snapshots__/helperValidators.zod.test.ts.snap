// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helperValidators emptyStringToUndefinedValidation accept empty string 1`] = `
{
  "data": undefined,
  "success": true,
}
`;

exports[`helperValidators emptyStringToUndefinedValidation accept undefined 1`] = `
{
  "data": undefined,
  "success": true,
}
`;

exports[`helperValidators emptyStringToUndefinedValidation decline filled string 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "message": "Invalid input",
    "path": []
  }
]],
  "success": false,
}
`;

exports[`helperValidators emptyStringToUndefinedValidation decline number 1`] = `
{
  "error": [ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [],
    "message": "Expected string, received number"
  }
]],
  "success": false,
}
`;

exports[`helperValidators getSingleFileValidation .docm 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "i18n": "errors.file_extension_invalid_type"
    },
    "path": [],
    "message": "Invalid input"
  }
]],
  "success": false,
}
`;

exports[`helperValidators getSingleFileValidation .tar.gz 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "i18n": "errors.file_extension_invalid_type"
    },
    "path": [],
    "message": "Invalid input"
  }
]],
  "success": false,
}
`;

exports[`helperValidators getSingleFileValidation .webm 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "i18n": "errors.file_extension_invalid_type"
    },
    "path": [],
    "message": "Invalid input"
  }
]],
  "success": false,
}
`;

exports[`helperValidators getSingleFileValidation Linux/UNIX hidden file 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "i18n": "errors.file_basename_too_small"
    },
    "path": [],
    "message": "Invalid input"
  },
  {
    "code": "custom",
    "params": {
      "i18n": "errors.file_extension_invalid_type"
    },
    "path": [],
    "message": "Invalid input"
  }
]],
  "success": false,
}
`;

exports[`helperValidators getSingleFileValidation crazy characters file 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "i18n": "errors.file_name_invalid_characters"
    },
    "path": [],
    "message": "Invalid input"
  }
]],
  "success": false,
}
`;

exports[`helperValidators getSingleFileValidation no basename file 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "i18n": "errors.file_basename_too_small"
    },
    "path": [],
    "message": "Invalid input"
  },
  {
    "code": "custom",
    "params": {
      "i18n": "errors.file_extension_invalid_type"
    },
    "path": [],
    "message": "Invalid input"
  }
]],
  "success": false,
}
`;

exports[`helperValidators getSingleFileValidation no prefix file 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "i18n": "errors.file_basename_too_small"
    },
    "path": [],
    "message": "Invalid input"
  },
  {
    "code": "custom",
    "params": {
      "i18n": "errors.file_extension_invalid_type"
    },
    "path": [],
    "message": "Invalid input"
  }
]],
  "success": false,
}
`;

exports[`helperValidators getSingleFileValidation standard file 1`] = `
{
  "data": {
    "fileName": "awoo.png",
    "size": 120,
  },
  "success": true,
}
`;

exports[`helperValidators getSingleFileValidation very big sized file 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "i18n": "errors.file_size_too_large",
      "size": 100000
    },
    "path": [],
    "message": "Invalid input"
  }
]],
  "success": false,
}
`;

exports[`helperValidators getSingleFileValidation zero sized file 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "i18n": "errors.file_size_too_small"
    },
    "path": [],
    "message": "Invalid input"
  }
]],
  "success": false,
}
`;

exports[`helperValidators partnerIdValidation accept project participant id 1`] = `
{
  "data": "a0D2600000zXrAIEA0",
  "success": true,
}
`;

exports[`helperValidators partnerIdValidation decline account id 1`] = `
{
  "error": [ZodError: [
  {
    "code": "invalid_string",
    "validation": {
      "startsWith": "a0D"
    },
    "message": "Invalid input: must start with \\"a0D\\"",
    "path": []
  }
]],
  "success": false,
}
`;

exports[`helperValidators partnerIdValidation decline empty string 1`] = `
{
  "error": [ZodError: [
  {
    "code": "invalid_string",
    "validation": {
      "startsWith": "a0D"
    },
    "message": "Invalid input: must start with \\"a0D\\"",
    "path": []
  }
]],
  "success": false,
}
`;

exports[`helperValidators partnerIdValidation decline project id 1`] = `
{
  "error": [ZodError: [
  {
    "code": "invalid_string",
    "validation": {
      "startsWith": "a0D"
    },
    "message": "Invalid input: must start with \\"a0D\\"",
    "path": []
  }
]],
  "success": false,
}
`;

exports[`helperValidators periodIdValidation accept 1 1`] = `
{
  "data": 1,
  "success": true,
}
`;

exports[`helperValidators periodIdValidation accept 2 1`] = `
{
  "data": 2,
  "success": true,
}
`;

exports[`helperValidators periodIdValidation accept 3 1`] = `
{
  "data": 3,
  "success": true,
}
`;

exports[`helperValidators periodIdValidation accept 5 1`] = `
{
  "data": 5,
  "success": true,
}
`;

exports[`helperValidators periodIdValidation accept string int 1`] = `
{
  "data": 4,
  "success": true,
}
`;

exports[`helperValidators periodIdValidation decline 0 1`] = `
{
  "error": [ZodError: [
  {
    "code": "too_small",
    "minimum": 0,
    "type": "number",
    "inclusive": false,
    "exact": false,
    "message": "Number must be greater than 0",
    "path": []
  }
]],
  "success": false,
}
`;

exports[`helperValidators periodIdValidation decline absurd number 1`] = `
{
  "error": [ZodError: [
  {
    "code": "too_big",
    "maximum": 500,
    "type": "number",
    "inclusive": false,
    "exact": false,
    "message": "Number must be less than 500",
    "path": []
  }
]],
  "success": false,
}
`;

exports[`helperValidators periodIdValidation decline decimal 1`] = `
{
  "error": [ZodError: [
  {
    "code": "invalid_type",
    "expected": "integer",
    "received": "float",
    "message": "Expected integer, received float",
    "path": []
  }
]],
  "success": false,
}
`;

exports[`helperValidators periodIdValidation decline infinity 1`] = `
{
  "error": [ZodError: [
  {
    "code": "invalid_type",
    "expected": "integer",
    "received": "float",
    "message": "Expected integer, received float",
    "path": []
  },
  {
    "code": "too_big",
    "maximum": 500,
    "type": "number",
    "inclusive": false,
    "exact": false,
    "message": "Number must be less than 500",
    "path": []
  }
]],
  "success": false,
}
`;

exports[`helperValidators projectIdValidation accept project id 1`] = `
{
  "data": "a0E2600000olswLEAQ",
  "success": true,
}
`;

exports[`helperValidators projectIdValidation decline account id 1`] = `
{
  "error": [ZodError: [
  {
    "code": "invalid_string",
    "validation": {
      "startsWith": "a0E"
    },
    "message": "Invalid input: must start with \\"a0E\\"",
    "path": []
  }
]],
  "success": false,
}
`;

exports[`helperValidators projectIdValidation decline empty string 1`] = `
{
  "error": [ZodError: [
  {
    "code": "invalid_string",
    "validation": {
      "startsWith": "a0E"
    },
    "message": "Invalid input: must start with \\"a0E\\"",
    "path": []
  }
]],
  "success": false,
}
`;

exports[`helperValidators projectIdValidation decline project participant id 1`] = `
{
  "error": [ZodError: [
  {
    "code": "invalid_string",
    "validation": {
      "startsWith": "a0E"
    },
    "message": "Invalid input: must start with \\"a0E\\"",
    "path": []
  }
]],
  "success": false,
}
`;
